firewood/src/merkle/node.rs

```diff
@@ -459,6 +463,203 @@ impl Storable for Node {
+    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        use serde::ser::Error;
+
+        match &self.node {
+            EncodedNodeType::Leaf(n) => {
+                let list = [
+                    Encoded::Raw(from_nibbles(&n.path.encode(true)).collect()),
+                    Encoded::Raw(n.data.to_vec()),
+                ];
+                let mut seq = serializer.serialize_seq(Some(list.len()))?;
+                for e in list {
+                    seq.serialize_element(&e)?;
+                }
+                seq.end()
+            }
+            EncodedNodeType::Branch { children, value } => {
+                let mut list = <[Encoded<Vec<u8>>; BranchNode::MAX_CHILDREN + 1]>::default();
+
+                for (i, c) in children.iter().enumerate() {
+                    if let Some(c) = c {
+                        if c.len() >= TRIE_HASH_LEN {
+                            let serialized_hash =
+                                Bincode::serialize(&Keccak256::digest(c).to_vec())
+                                    .map_err(|e| S::Error::custom(format!("bincode error: {e}")))?;
+                            list[i] = Encoded::Data(serialized_hash);
+                        } else {
+                            list[i] = Encoded::Raw(c.to_vec());
+                        }
+                    }
+                }
+                if let Some(Data(val)) = &value {
+                    let serialized_val = Bincode::serialize(val)
+                        .map_err(|e| S::Error::custom(format!("bincode error: {e}")))?;
+                    list[BranchNode::MAX_CHILDREN] = Encoded::Data(serialized_val);
+                }
+
+                let mut seq = serializer.serialize_seq(Some(list.len()))?;
+                for e in list {
+                    seq.serialize_element(&e)?;
+                }
+                seq.end()
+            }
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for EncodedNode<Bincode> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use serde::de::Error;
+
+        let items: Vec<Encoded<Vec<u8>>> = Deserialize::deserialize(deserializer)?;
+        let len = items.len();
+        match len {
+            LEAF_NODE_SIZE => {
+                let mut items = items.into_iter();
+                let Some(Encoded::Raw(path)) = items.next() else {
+                    return Err(D::Error::custom(
+                        "incorrect encoded type for leaf node path",
+                    ));
+                };
+                let Some(Encoded::Raw(data)) = items.next() else {
+                    return Err(D::Error::custom(
+                        "incorrect encoded type for leaf node data",
+                    ));
+                };
+                let path_nibbles = Nibbles::<0>::new(&path);
+                let path = PartialPath::from_nibbles(path_nibbles.into_iter()).0;
```

Comment:
nit: I would have combined these:
```suggestion
                let path = PartialPath::from_nibbles(Nibbles::<0>::new(&path).into_iter()).0;
```